### PR TITLE
[trainer] feat: diffusers FSDP engine support sequence parallel

### PR DESCRIPTION
**Pull request by zhtmike**
### What does this PR do?
- The `diffusers` FSDP engine supports sequence parallelism (require `diffusers` >= 0.38).  
- Add a `diffusers` test to check forward and backward accuracy. It is not included in the smoke test since we rarely change the `diffusers` version compared with `verl`. Might be in CI in future.
- We use `diffusers`’ own SP mesh instead of `verl`’s to avoid hacks or patches.
- There is no scaling in `diffusers`’ SP backward pass, so we need to manually scale the loss by SP before performing backward, may check the evidence in the SP UT.
- We need to add a patch for `QwenImageTransformer2DModel`, since the mask is not sharded correctly under Ulysses SP, which causes `ppo_kl` and `pg_clipfrac_higher` to have a significant bias. SP UT can reproduce this bug by dropping this patch.

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

Exps with SP=2 with SP=1 (no SP)

Actor loss and `ppo_kl` is in same scale w/wo SP:
螢幕截圖 2026-05-08 上午9 52 11

Critic reward is in consistent:
螢幕截圖 2026-05-08 上午9 55 57

Validation reward:
螢幕截圖 2026-05-08 上午9 56 12

Memory peak for SP is dropped compared with non-SP
螢幕截圖 2026-05-08 上午9 57 09

Time is increased since the DP is effectively halved
螢幕截圖 2026-05-08 上午9 57 35

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update the documentation.
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

File: docs/algo/flowgrpo.md [hunk 1]
```diff
@@ -108,6 +108,29 @@ the rollout section is the main place to override sampling behavior.
...
+```
+dp_size = total_gpus / ulysses_sequence_parallel_size
+```
+
+For SP training, `num_attention_heads` must be divisible by
+`ulysses_sequence_parallel_size`.
+
+A ready-to-use 4-GPU SP=2 example is provided:
+```bash
+bash examples/flowgrpo_trainer/run_qwen_image_ocr_lora_sp2.sh
```
Comment (AndyZhou952) [inline docs/algo/flowgrpo.md]: Maybe put this under variants since SP is not unique to flowgrpo?

File: tests/workers/test_diffusers_ulysses.py [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+        output_sp = module_sp(**model_inputs)[0]
+
+    # 2. plain (non-SP) forward
+    module_no_sp.eval()
+    with torch.no_grad():
+        output_no_sp = module_no_sp(**model_inputs)[0]
+
+    assert output_sp.shape == output_no_sp.shape, f"Shape mismatch: SP {output_sp.shape} vs non-SP {output_no_sp.shape}"
+
+    torch.testing.assert_close(output_sp.float(), output_no_sp.float(), rtol=1e-2, atol=1e-2)
```
Comment (AndyZhou952) [inline tests/workers/test_diffusers_ulysses.py:192]: Maybe a note here to prevent future AI from violently modifying the threshold to pass UT?

**Review by AndyZhou952 — APPROVED** [review]

File: docs/algo/flowgrpo.md (hunk 1 above)
Comment (AndyZhou952) [reply docs/algo/flowgrpo.md]: For clarity: the variants section in this doc for now.

Comment (zhtmike) [reply docs/algo/flowgrpo.md]: done

File: tests/workers/test_diffusers_ulysses.py (hunk 2 above)
Comment (zhtmike) [reply tests/workers/test_diffusers_ulysses.py:192]: I think it is better to put it in a general place like AGENT.md ?

Btw I expect the reviewer should takes responsibility to guard the commit into the main

Comment (AndyZhou952) [reply tests/workers/test_diffusers_ulysses.py:192]: Agreed on developing a general rule later. For this one, it’s worth mentioning the bias issue that influenced the threshold. In other cases, flaky CI might make a looser threshold more appropriate so probably need to decide on a case-by-case basis.

Comment (zhtmike) [reply tests/workers/test_diffusers_ulysses.py:192]: Agree. Updated
